### PR TITLE
Bind to 0.0.0.0

### PR DESCRIPTION
**Pull request by asonnino**
Binding to our external ip address causes problems when running on AWS. That is, running the following command immediately crashes the validator:
`cargo run --release --bin sui-node -- --config-path ~/.sui/sui_config/validator-config-3.yaml `

```
ubuntu@ip-172-31-0-27:~/sui$ cargo run --release --bin sui-node -- --config-path ~/.sui/sui_config/validator-config-3.yaml 
    Finished release [optimized + debuginfo] target(s) in 0.80s
     Running `target/release/sui-node --config-path /home/ubuntu/.sui/sui_config/validator-config-3.yaml`
2023-02-28T10:56:57.674364Z  INFO sui_node: Sui Node version: 0.28.0-4cc5d68f5
2023-02-28T10:56:57.674388Z  INFO sui_node: Supported protocol versions: Some(SupportedProtocolVersions { min: ProtocolVersion(1), max: ProtocolVersion(1) })
2023-02-28T10:56:57.674396Z  INFO sui_node: Started Prometheus HTTP endpoint at 127.0.0.1:45593
2023-02-28T10:56:57.674429Z  INFO sui_node::admin: starting admin server filter=info address=127.0.0.1:37553
2023-02-28T10:56:57.674466Z  INFO sui_node: Initializing sui-node listening on /ip4/3.81.207.116/tcp/1500/http node=k#898036fb4c6928c7446928b9524a37649b08d54428284d5e23f99decf0a298c43c552849c2393c9b29d43242171bed5b060739739471435ecb2f9378772e2e8c68c86f224dfcca8b230df8a7262f06afcb1b72add0c43fc3d981b73de3919631
2023-02-28T10:56:58.071107Z  INFO typed_store::rocks: Returning the cf metric logging task for DBMap: checkpoint_content
2023-02-28T10:56:58.071147Z  INFO typed_store::rocks: Returning the cf metric logging task for DBMap: certified_checkpoints
2023-02-28T10:56:58.071151Z  INFO typed_store::rocks: Returning the cf metric logging task for DBMap: epoch_last_checkpoint_map
2023-02-28T10:56:58.071154Z  INFO typed_store::rocks: Returning the cf metric logging task for DBMap: watermarks
2023-02-28T10:56:58.071157Z  INFO typed_store::rocks: Returning the cf metric logging task for DBMap: transactions
2023-02-28T10:56:58.071161Z  INFO typed_store::rocks: Returning the cf metric logging task for DBMap: effects_signatures
2023-02-28T10:56:58.071164Z  INFO typed_store::rocks: Returning the cf metric logging task for DBMap: transaction_cert_signatures
2023-02-28T10:56:58.071167Z  INFO typed_store::rocks: Returning the cf metric logging task for DBMap: assigned_shared_object_versions
2023-02-28T10:56:58.071171Z  INFO typed_store::rocks: Returning the cf metric logging task for DBMap: next_shared_object_versions
2023-02-28T10:56:58.071174Z  INFO typed_store::rocks: Returning the cf metric logging task for DBMap: pending_execution
2023-02-28T10:56:58.071177Z  INFO typed_store::rocks: Returning the cf metric logging task for DBMap: consensus_message_processed
2023-02-28T10:56:58.071180Z  INFO typed_store::rocks: Returning the cf metric logging task for DBMap: pending_consensus_transactions
2023-02-28T10:56:58.071183Z  INFO typed_store::rocks: Returning the cf metric logging task for DBMap: consensus_message_order
2023-02-28T10:56:58.071186Z  INFO typed_store::rocks: Returning the cf metric logging task for DBMap: last_consensus_index
2023-02-28T10:56:58.071189Z  INFO typed_store::rocks: Returning the cf metric logging task for DBMap: checkpoint_boundary
2023-02-28T10:56:58.071192Z  INFO typed_store::rocks: Returning the cf metric logging task for DBMap: reconfig_state
2023-02-28T10:56:58.071181Z  INFO typed_store::rocks: Returning the cf metric logging task for DBMap: checkpoint_by_digest
2023-02-28T10:56:58.071196Z  INFO typed_store::rocks: Returning the cf metric logging task for DBMap: end_of_publish
2023-02-28T10:56:58.071200Z  INFO typed_store::rocks: Returning the cf metric logging task for DBMap: final_epoch_checkpoint
2023-02-28T10:56:58.071204Z  INFO typed_store::rocks: Returning the cf metric logging task for DBMap: pending_checkpoints
2023-02-28T10:56:58.071208Z  INFO typed_store::rocks: Returning the cf metric logging task for DBMap: builder_digest_to_checkpoint
2023-02-28T10:56:58.071211Z  INFO typed_store::rocks: Returning the cf metric logging task for DBMap: pending_checkpoint_signatures
2023-02-28T10:56:58.071214Z  INFO typed_store::rocks: Returning the cf metric logging task for DBMap: user_signatures_for_checkpoints
2023-02-28T10:56:58.071219Z  INFO typed_store::rocks: Returning the cf metric logging task for DBMap: builder_checkpoint_summary
2023-02-28T10:56:58.071224Z  INFO typed_store::rocks: Returning the cf metric logging task for DBMap: epoch_start_configuration
2023-02-28T10:56:58.071229Z  INFO typed_store::rocks: Returning the cf metric logging task for DBMap: state_hash_by_checkpoint
2023-02-28T10:56:58.071233Z  INFO typed_store::rocks: Returning the cf metric logging task for DBMap: authority_capabilities
2023-02-28T10:56:58.071248Z  INFO typed_store::rocks: Returning the cf metric logging task for DBMap: log
2023-02-28T10:56:58.071252Z  INFO typed_store::rocks: Returning the cf metric logging task for DBMap: retry_count
2023-02-28T10:56:58.071251Z  INFO typed_store::rocks: Returning the cf metric logging task for DBMap: num_certified_checkpoint_signatures
2023-02-28T10:56:58.071330Z  INFO typed_store::rocks: Returning the cf metric logging task for DBMap: objects
2023-02-28T10:56:58.071358Z  INFO typed_store::rocks: Returning the cf metric logging task for DBMap: owned_object_transaction_locks
2023-02-28T10:56:58.071365Z  INFO typed_store::rocks: Returning the cf metric logging task for DBMap: transactions
2023-02-28T10:56:58.071370Z  INFO typed_store::rocks: Returning the cf metric logging task for DBMap: effects
2023-02-28T10:56:58.071375Z  INFO typed_store::rocks: Returning the cf metric logging task for DBMap: executed_effects
2023-02-28T10:56:58.071381Z  INFO typed_store::rocks: Returning the cf metric logging task for DBMap: executed_transactions_to_checkpoint
2023-02-28T10:56:58.071386Z  INFO typed_store::rocks: Returning the cf metric logging task for DBMap: root_state_hash_by_epoch
2023-02-28T10:56:58.071392Z  INFO typed_store::rocks: Returning the cf metric logging task for DBMap: current_epoch
2023-02-28T10:56:58.071403Z  INFO typed_store::rocks: Returning the cf metric logging task for DBMap: parent_sync
2023-02-28T10:56:58.071499Z  INFO typed_store::rocks: Returning the cf metric logging task for DBMap: committee_map
Error: Cannot assign requested address (os error 99)
```

### CODE REVIEW

File: crates/sui-config/src/node.rs [hunk 1]
```diff
@@ -101,6 +101,34 @@ pub struct NodeConfig {
     pub supported_protocol_versions: Option<SupportedProtocolVersions>,
 }
 
+impl NodeConfig {
+    /// Zero the ip address to ensure we bind to `0.0.0.0`.
+    pub fn zero_ip_address(&mut self) {
```
Comment (akichidis) [inline crates/sui-config/src/node.rs:106]: I would try to give a bit more specific name to this method, ex `zero_network_adresses`

File: crates/sui-config/src/node.rs [hunk 2]
```diff
@@ -101,6 +101,34 @@ pub struct NodeConfig {
...
+            .set_ip("0.0.0.0".parse().unwrap());
+
+        // Set the ip network address to `0.0.0.0`.
+        self.network_address = Self::zero_ip_multi_address(&self.network_address);
+    }
+
+    /// Set the ip address to `0.0.0.0`. For instance, it converts the following address
+    /// `/ip4/155.138.174.208/tcp/1500/http` into `/ip4/0.0.0.0/tcp/1500/http`.
+    fn zero_ip_multi_address(address: &Multiaddr) -> Multiaddr {
+        let mut new_address = Multiaddr::empty();
```
Comment (akichidis) [inline crates/sui-config/src/node.rs:119]: What the empty address will resolve to?

**Comment by mystenmark** [comment]
LGTM but I think @bmwill and/or @aschran should take a quick look

**Review by bmwill — CHANGES_REQUESTED** [review]
Please lets not do this. The Node config already has support for configuring the listen addresses for these various different network interfaces. Please use the existing configurations and set them appropriately. Doing this would instead override any user provided configuration leading to very confused node operators.

**Comment by asonnino** [comment]
> Please lets not do this. The Node config already has support for configuring the listen addresses for these various different network interfaces. Please use the existing configurations and set them appropriately. Doing this would instead override any user provided configuration leading to very confused node operators.

I see your point, I missed the option `--listen-address`. How about this change instead @bmwill ? It seems we also need to adapt the p2p listen address

File: crates/sui-config/src/node.rs [hunk 3]
```diff
@@ -101,6 +101,34 @@ pub struct NodeConfig {
     pub supported_protocol_versions: Option<SupportedProtocolVersions>,
 }
 
+impl NodeConfig {
+    /// Zero the ip address to ensure we bind to `0.0.0.0`.
+    pub fn zero_ip_address(&mut self) {
+        // Set the p2p2 listen address to `0.0.0.0`.
+        self.p2p_config
+            .listen_address
```
[the remark this answers is not in the document]
Comment (asonnino) [reply crates/sui-config/src/node.rs]: mmh I See so we can just set the p2p listen address to 0.0.0.0 in the configs and the client should read the external address

**Comment by bmwill** [comment]
@asonnino No, the change you put up is also not needed. Both the grpc endpoint as well as the p2p one can be fully configured via the NodeConfig. 

https://github.com/MystenLabs/sui/blob/main/crates/sui-config/src/node.rs#L48 for the grpc endpoint and
https://github.com/MystenLabs/sui/blob/main/crates/sui-config/src/p2p.rs#L14 for the p2p endpoint. If you're having issues with the listen addesses on node startup please change and adjust what is provided in the configs themselves.

**Comment by asonnino** [comment]
> @asonnino No, the change you put up is also not needed. Both the grpc endpoint as well as the p2p one can be fully configured via the NodeConfig.
> 
> https://github.com/MystenLabs/sui/blob/main/crates/sui-config/src/node.rs#L48 for the grpc endpoint and https://github.com/MystenLabs/sui/blob/main/crates/sui-config/src/p2p.rs#L14 for the p2p endpoint. If you're having issues with the listen addesses on node startup please change and adjust what is provided in the configs themselves.

@bmwill 
Should I then modify the genesis config and the genesis builder to expose this? It seems the current genesis builder sets the listen address to the same value as the external address 

https://github.com/MystenLabs/sui/blob/00bb9bb66f0315e4ae87210ac1da4a19d30e9dfd/crates/sui-config/src/builder.rs#L395-L402

**Comment by bmwill** [comment]
@asonnino what is the use case you're trying to solve for? If this is for deploying to a large cluster of nodes then it may be easier to either add a step in that deployment to do this config modification. Otherwise, yes adding a parameter to that genesis configuration to optionally bind to 0.0.0.0 vs what its doing today would be an option. But i wouldn't change that process to unconditionally bind to 0.0.0.0 because we have tests that i think we should continue requiring that they bind to 127.0.0.1 as they do today

**Comment by asonnino** [comment]
> @asonnino what is the use case you're trying to solve for? If this is for deploying to a large cluster of nodes then it may be easier to either add a step in that deployment to do this config modification. Otherwise, yes adding a parameter to that genesis configuration to optionally bind to 0.0.0.0 vs what its doing today would be an option. But i wouldn't change that process to unconditionally bind to 0.0.0.0 because we have tests that i think we should continue requiring that they bind to 127.0.0.1 as they do today

I am trying to deploy small (but many) clusters of nodes on AWS for benchmarking but it seems AWS cannot bind to the external address. They are no large PE deployments so I am trying keep the overhead small. I can see if we can specify the listening address through the genesis config with minimal changes to our current codebase

**Pull request closed without merging**